ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -0,0 +1,777 @@
+    --     i.e. we switch to a short fork.
+    --   - In exceptional cases, it will be exactly @k@ long, when we switch
+    --     to a fork of the same length.
+    --   - In exceptional cases, it will be much longer than @k@, i.e. when we
+    --     have added a block that connects a much longer fork to the current
+    --     chain.
+    --
+    -- * The current chain might be longer than @k@ because the background
+    --   thread might not have written all blocks yet to the ImmutableDB.
+    --
+    -- If there's no concurrency, i.e., no concurrent calls to 'addBlock',
+    -- then it must be that current chain and the candidate chain intersect
+    -- within the last @k@ blocks (of the current chain).
+    --
+    -- If there is no such intersection, it must be that the current chain has
+    -- /concurrently/ changed so much that the fragment is no longer within
+    -- the last @k@ blocks of the current chain, and we are no longer
+    -- interested in this candidate anymore.
+    --
+    -- If there is an intersection, we compare the candidate fragment to the
+    -- current chain, and, if it is preferred, install it as the current
+    -- chain.
+    --
+    -- In case the current chain was changed in the background, it must either
+    -- have been done before we added our block to the VolatileDB or after. If
+    -- it was before, we know of the other block. If it was done after, they
+    -- must know of our block. In either case, somebody computed candidates
+    -- with knowledge of both blocks, so we're safe. See See "### Concurrency"
+    -- in ChainDB.md for more details.
+    trySwitchTo :: HasCallStack
+                => ChainAndLedger blk  -- ^ Chain and ledger to switch to
+                -> m ()
+    trySwitchTo (ChainAndLedger newChain newLedger) = do
+      (curChain, switched) <- atomically $ do
+        curChain <- readTVar cdbChain
+        case AF.intersect curChain newChain of
+          Just (curChainPrefix, _newChainPrefix, curChainSuffix, newChainSuffix)
+            | fromIntegral (AF.length curChainSuffix) <= k
+            , preferCandidate cdbNodeConfig curChain newChain
+            -> do
+              -- The current chain might still contain some old headers that
+              -- have not yet been written to the ImmutableDB. We must make
+              -- sure to prepend these headers to the new chain so that we
+              -- still know what the full chain is.
+              let newChain' = fromMaybe
+                    (error "postcondition of intersect violated") $
+                    AF.join curChainPrefix newChainSuffix
+              writeTVar cdbChain newChain'
+              LgrDB.setCurrent cdbLgrDB newLedger
+              -- Update 'cdbImmBlockNo'
+              modifyTVar' cdbImmBlockNo $ getImmBlockNo secParam newChain'
+
+              -- Update the readers
+              let ipoint = castPoint $ AF.anchorPoint newChainSuffix
```

Comment:
Comment about the precondition of `switchFork` and that we're satisfying it here.